src/cachingreader.cpp

```diff
@@ -162,66 +159,90 @@ void CachingReader::freeAllChunks() {
     m_allocatedChunks.clear();
     m_mruChunk = NULL;
 
-    QSet<Chunk*> reserved = QSet<Chunk*>::fromList(m_chunksBeingRead.values());
+    for (QVector<Chunk*>::const_iterator it = m_chunks.constBegin();
+         it != m_chunks.constEnd(); ++it) {
+        Chunk* pChunk = *it;
 
-    for (int i=0; i < m_chunks.size(); i++) {
-        Chunk* c = m_chunks[i];
-        if (reserved.contains(c)) {
+        // We will receive CHUNK_READ_INVALID for all pending chunk reads which
+        // should free the chunks individually.
+        if (pChunk->state == Chunk::READ_IN_PROGRESS) {
             continue;
         }
-        if (!m_freeChunks.contains(c)) {
-            c->chunk_number = -1;
-            c->length = 0;
-            c->next_lru = NULL;
-            c->prev_lru = NULL;
-            m_freeChunks.push_back(c);
+
+        if (pChunk->state != Chunk::FREE) {
```

Comment:
I didn't profile this since I was mostly profiling a single track in the steady state playing.